# Genome Analysis and Quality Check Pipeline
This repository contains the data processing workflow developed as part of our 7th semester project, which aims to characterise soil-derived bacterial isolates using Oxford Nanopore sequencing and explore their potential for antibiotic production.

The workflow systematically processes raw Nanopore reads through filtering, quality assessment, genome assembly, polishing, and multiple layers of assembly quality evaluation, including completeness and contamination checks. Finally, it performs gene annotation and taxonomic classification to provide a detailed genomic overview.

---

## Repository Contents
The repository contains both a full end-to-end pipeline and modular scripts that allow individual processing steps to be executed independently.

### Main pipeline scripts
- `Pipeline.sh`: A pipeline script that automates the analysis workflow from raw Nanopore FASTQ files through to assembled polished genomes and QC.
- `GTDB-Tk.sh`: Script for taxonomic assignment using GTDB-Tk.
- `Bakta.sh`: Script for genome annotation.



### Modular scripts
The following modular scripts enable individual pipeline steps to be run separately for testing, re-analysis, or parameter tuning:
- `1. Filtering`
- `2.1. NanoPlot on raw reads`
- `2.2. NanoPlot on filtered reads`
- `3. Assembly with Flye`
- `4. Coverage`
- `5. Polishing with Medaka`
- `6. QUAST`
- `7. CheckM`
- `8. CheckM2`

Important note: Most modular scripts depend on outputs generated by one or more preceding steps (e.g. filtered reads, assemblies, or polished genomes). Therefore, prerequisite steps must be completed before running a given module independently.

---

## Workflow Overview
### Diagram of Workflow
![alt text](https://github.com/huooong/Genome-Assembly-and-QC/blob/main/Documents/Bioinformatic%20workflow.png)

### Steps of the Workflow
**1. Read filtering using Filtlong**  
   - Filters raw Nanopore reads to improve downstream assembly quality using the following parameters:
      - Minimum read length: **1000 bp**,  
      - Keep top **90%** highest-quality reads,
      - Trim bases until only the best **500 Mbp** remain.

**2. Quality assessment of reads using NanoPlot**
   - Provides summary statistics and visualisations describing the quality of unfiltered and filtered reads.
   - Generated plots include:
      - Weighted and non-weighted histogram of read lengths,
      - Length vs. Quality scatter plot,
      - Yield by length plot.
        
**3. Genome assembly with Flye**
   - Assembles genomes from filtered Nanopore reads using the following parameters:
      - Estimated genome size: **5 Mb**.

**4. Coverage calculation**  
   - Maps filtered reads back to assemblies and calculates sequencing coverage using Minimap2 and SAMtools.  

**5. Assembly polishing with Medaka**  
   - Refines assemblies using Medaka
      - Model: `r1041_e82_400bps_hac_v5.0.0`

**6. Assembly quality assessment with QUAST**
   - Evaluates assembly statistics including number of contigs, N50, contig length, and GC content.

**7. Assessment of completeness and contamination with CheckM and CheckM2**
   - CheckM estimates genome completeness and contamination based on the presense/absense of marker genes.
   - CheckM2 uses machine learning models to estimate completeness and contamination.

**Genome annotation using Bakta**
   - Annotates assembled genomes with gene features and functional assignments.

**Taxonomy assignment with GTDB-Tk**
   - Assigns taxonomy to assemblies based on the Genome Taxonomy Database.

---
## Directories 
### Project Structure

The main project directory is called `Working_Directory`. Within `Working_Directory`, there is a `Data` directory containing all raw FASTQ files (`*.fastq.gz`) generated from Oxford Nanopore sequencing. 

The fastq.gz files are named `PBI54877.barcodeX.fastq.gz`, where X = 17-36, corresponding to individual samples.

### Output Organisation
All analysis results are stored inside a folder called `Results`. This directory contains subfolders for each step of the pipeline: 
   - Filtered/
   - NanoPlot/
   - NanoPlot_filtered/
   - Assembly/
   - Coverage/
   - Medaka/
   - QUAST/
   - CheckM/
   - CheckM2/

Each of these subfolders contains one directory per sample, following the same naming scheme as the input files. 

---

## Setup and How to Run the Pipeline
1. Download/copy the `Pipeline.sh` and `Pipeline.sbatch` to `Working_Directory`

2. Prepare input files.
Place all raw FASTQ files in the directory `Working_Directory/Data`.
```bash
cp -r /raw_data/MA/tinyearth/20251119_np_PBI54877/*.fastq.gz ~/Working_Directory/Data/
```

3. Go to the terminal, log on, and change to your working directory:
```bash
cd Working_Directory
````

4. Run the pipeline:
```bash
sbatch Pipeline.sbatch
````

5. Status:
```bash
squeue -u $USER
```

```bash
tail -f logs/genome_assembly_and_QC_<jobID>.out
```

---
### Databases
For CheckM2, GTDB-Tk and Bakta, you must specify the path to the corresponding reference databases before running the scripts. These databases are not included in this repository and must be installed separately according to the official tool documentation.

Ensure that the database paths are correctly set in the respective scripts or provided as environment variables prior to execution.

---

### Conda Environment Activation Issue
When running scripts inside VS Code or Code-Server using the "Run Active File" feature, you might encounter an error like:
```
CondaError: Run 'conda init' before 'conda activate'
```
This happens because VS Code runs script in a non-interactive shell. 

To fix it, add:
```bash
CONDA_BASE=$(dirname $(dirname $(which conda)))
source "$CONDA_BASE/etc/profile.d/conda.sh"
```
Before activating conda environment

---

## Software Requirements
This pipeline relies on several bioinformatics tools and libraries. These versions were employed during analysis.
- NanoPlot (v1.46.1)
- Filtlong (v0.2.1)
- Flye (v2.9.6-b1802)
- Minimap2 (v2.30-r1287)
- SAMtools (v1.22.1)
- QUAST (v5.3.0)
- Medaka (v2.1.1)
- CheckM1 (v1.2.4)
- CheckM2 (v1.1.0) and database (uniref100.KO.1.dmnd)
- Bakta (v1.11.4) and database (v6.0, 2025-02-24)
- GTDB-Tk (v2.5.2) and database (release 226, 2025-04-11)
  
---

## Resources
**Filtlong**:
- Github: https://github.com/rrwick/Filtlong 

**Nanoplot**:
- Github: https://github.com/wdecoster/NanoPlot
- Article:
   - De Coster et al. (2018). NanoPack: visualizing and processing long-read sequencing data. _Bioinformatics_, 34 (15): 2666-2669, (2018)
  
**Flye**:
- Github: https://github.com/mikolmogorov/Flye
- Articles:
   - **Single genome assembly:** Kolmogorov et al. (2019). Assembly of long, error-prone reads using repeat graphs. _Nature Biotechnology_, 37 (2019) 540-546.
   - **Core algorithm:** Lin et al. (2016). Assembly of long error-prone reads using de Bruijn graphs. _PNAS_, 113 (52) E8396-E8405.

**Minimap2**
- Github: https://github.com/lh3/minimap2
- Article:
   - Li, H. (2018). Minimap2: pairwise alignment for nucleotide sequences. _Bioinformatics_, 34(18):3094–3100.    
  
**SAMtools**
- Github: https://github.com/samtools/samtools
- Article:
   - Danecek et al. (2021). Twelve years of SAMtools and BCFtools. _GigaScience_, 10 (2).
  
**Medaka**: 
- Github: https://github.com/nanoporetech/medaka

**QUAST**:
- Github: https://github.com/ablab/quast
- Article:
   - Gurevich et al. (2013). QUAST: quality assessment tool for genome assemblies. _Bioinformatics_, 29(8):1072–1075.
     
**CheckM**:
- Github: https://github.com/Ecogenomics/CheckM
- Article:
   - Parks et al. (2015). CheckM: assessing the quality of microbial genomes recovered from isolates, single cells,
and metagenomes. _Genome Research_, 25(7):1043–1055.
  
**CheckM2**:
- Github: https://github.com/chklovski/CheckM2/
- Article:
   - Chklovski et al. (2023). CheckM2: a rapid,
scalable and accurate tool for assessing microbial genome quality using machine learning. _Nature Methods_, 20:1203–1212.

**Bakta**:
- Github: https://github.com/oschwengers/bakta
- Article:
   - Schwengers et al. (2021). Bakta: rapid and standardized annotation of bacterial genomes via alignment-free sequence identification. Microbial Genomics, 7(11):000685.

**GTDB-Tk**:
- Github: https://ecogenomics.github.io/GTDBTk/index.html
- Article:
   - Chaumeil et al. (2020).  GTDB-Tk: a toolkit to
classify genomes with the Genome Taxonomy Database. _Bioinformatics_, 36(6):1925–192.
  
---
